python/ray/experimental/dag/input_node.py

```diff
@@ -43,22 +48,127 @@ def _copy_impl(
+            return args[0]
+
+        return DAGInputData(*args, **kwargs)
 
     def __str__(self) -> str:
         return get_dag_node_str(self, "__InputNode__")
 
+    def __getattr__(self, key: str):
+        return InputAtrributeNode(self, key)
+
+    def __getitem__(self, key: Union[int, str]) -> Any:
+        return InputAtrributeNode(self, key)
+
+    def __enter__(self):
+        # TODO (jiaodong): Supporting input node as context manager is easy.
+        # real question is we need to decide if we want to enforce that's the
+        # only way of creating input node.
+        return self
+
+    def __exit__(self, *args):
+        pass
+
     def to_json(self, encoder_cls) -> Dict[str, Any]:
-        # TODO: (jiaodong) Support arbitrary InputNode args and pydantic
-        # input schema.
         json_dict = super().to_json_base(encoder_cls, InputNode.__name__)
         return json_dict
 
     @classmethod
     def from_json(cls, input_json):
         assert input_json[DAGNODE_TYPE_KEY] == InputNode.__name__
-        # TODO: (jiaodong) Support user passing inputs to InputNode in JSON
         return cls()
+
+
+class InputAtrributeNode(DAGNode):
+    """Represents partial acces of user input based on an attribute key.
+
+    Examples:
+        >>> with InputNode() as dag_input:
+        >>>     a = input[0]
+        >>>     b = input.x
+
+        >>> # This makes a = 1 and b = 2
+        >>> ray_dag.execute(1, x=2)
+    """
+
+    def __init__(self, dag_input_node: InputNode, key: str):
+        self._dag_input_node = dag_input_node
+        self._key = key
+        super().__init__([], {}, {}, {"dag_input_node": dag_input_node, "key": key})
+
+    def _copy_impl(
+        self,
+        new_args: List[Any],
+        new_kwargs: Dict[str, Any],
+        new_options: Dict[str, Any],
+        new_other_args_to_resolve: Dict[str, Any],
+    ):
+        return InputAtrributeNode(
+            new_other_args_to_resolve["dag_input_node"],
+            new_other_args_to_resolve["key"],
+        )
+
+    def _execute_impl(self, *args, **kwargs):
+        """Executor of InputAtrributeNode.
+
+        Args and kwargs are to match base class signature, but not in the
+        implementation. All args and kwargs should be resolved and replaced
+        with value in bound_args and bound_kwargs via bottom-up recursion when
+        current node is executed.
+        """
+        if isinstance(self._dag_input_node, DAGInputData):
+            return self._dag_input_node[self._key]
+        else:
+            # dag.execute() is called with only one arg, thus when an
+            # InputAtrributeNode is executed, its dependent InputNode is
+            # resolved with original user input value.
+            return self._dag_input_node
+
+    def __str__(self) -> str:
+        return get_dag_node_str(self, f"__InputNode__[{self._key}]")
+
+
+class DAGInputData:
+    """Wrapped all user inputs as one object, accessible via attribute key.
```

Comment:
this is not actually the case, we don't support accessing via attribute right now (I think we should!)

Reply:
hmm i think we're supporting it already in this PR, but not in this object.

```
def __getattr__(self, key: str):
    return InputAtrributeNode(self, key)
```

This gives us a `InputAtrributeNode` upon calling `input.x` that turned to a dict access ;)

Reply:
updated -- new tests should make it very clear now.